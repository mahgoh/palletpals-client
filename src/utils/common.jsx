export function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

export function productImageURL(imageId) {
  return `${GLOBAL.API_URL}/product-images/${imageId}`
}

export function parseDateTime(dateTimeString, locale = 'en') {
  const datetime = new Date(dateTimeString)
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: 'numeric',
    minute: '2-digit',
  }

  return datetime.toLocaleString(locale, options)
}

export function formatPrice(price) {
  return `CHF ${price.toFixed(2)}`
}

export function formatOrderId(id) {
  return id.toString().padStart(6, '0')
}
