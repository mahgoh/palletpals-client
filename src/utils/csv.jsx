export async function processCSV(file) {
  try {
    const csv = await readCSV(file)
    const parsed = parseCSV(csv)

    return parsed
  } catch (e) {
    throw e
  }
}

function readCSV(file) {
  return new Promise((resolve, reject) => {
    if (file === null) reject('No file selected')

    // Check if the file is an image.
    if (file.type && !file.type === 'text/csv') {
      console.log('File is not csv.', file.type, file)
      reject('File is not csv.')
    }

    const reader = new FileReader()
    reader.addEventListener('load', (event) => {
      resolve(event.target.result)
    })
    reader.readAsText(file)
  })
}

/**
 *
 * Target:
 * {
 *   "km": [10,20,30],
 *   "pallets": [1,2,3],
 *   "price": [[5,6,7],[9,11,13],[15,17,19]]
 * }
 * @param {string} csv
 * @returns
 */

function parseCSV(csv) {
  const lines = csv.split('\n')

  // how entries are separated
  const delimiter = csv.includes(',') ? ',' : ';'

  const head = lines.shift()

  // palletArray is the first row except the first column
  // integers
  const palletArray = head
    .split(delimiter)
    .slice(1)
    .map((p) => parseInt(p))

  // kmArray is the first column except the first row
  // integers
  const kmArray = lines.map(
    (line) => line.split(delimiter).map((km) => parseInt(km))[0]
  )

  // priceMatrix is the rest of the table as two dimensional array
  // floats
  const priceMatrix = lines.map((line) =>
    line
      .split(delimiter)
      .slice(1)
      .map((p) => parseFloat(p))
  )

  return {
    kmArray,
    palletArray,
    priceMatrix,
  }
}
