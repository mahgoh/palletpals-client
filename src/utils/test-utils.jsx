import { cloneElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'
import i18n from '../i18n'

import { AuthProvider } from '@/services/auth'
import { AppearanceProvider } from '@/services/appearance'
import { CartProvider } from '@/services/cart'
import { LanguageProvider } from '@/services/language'
import { NotificationProvider } from '@/services/notification'

/**
 * Render component with router wrapper and option for specifying the current route.
 *
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouter = (component, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    ...render(component, { wrapper: BrowserRouter }),
  }
}

/**
 * Render component with router wrapper and i18next provider.
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouterWithi18n = (component, options) => {
  const comp = cloneElement(component, {
    changeLanguage: (lng) => {
      i18n.changeLanguage(lng)
      rerender(<I18nextProvider i18n={i18n}>{comp}</I18nextProvider>)
    },
  })
  const defaultRender = renderWithRouter(
    <I18nextProvider i18n={i18n}>{comp}</I18nextProvider>,
    options
  )
  const { rerender } = defaultRender
  return defaultRender
}

const renderComplete = (component, options) => {
  return {
    ...renderWithRouterWithi18n(
      <AuthProvider>
        <AppearanceProvider>
          <LanguageProvider>
            <CartProvider>
              <NotificationProvider>{component}</NotificationProvider>
            </CartProvider>
          </LanguageProvider>
        </AppearanceProvider>
      </AuthProvider>,
      options
    ),
  }
}

/**
 * Render component with specified renderer
 *
 * @param {JSX} component
 * @param {Object} param1
 * @returns
 */
const customRender = (component, { renderer = 'default', ...args } = {}) => {
  switch (renderer) {
    case 'router':
      return renderWithRouter(component, { ...args })
    case 'router-i18n':
      return renderWithRouterWithi18n(component, { ...args })
    case 'complete':
      return renderComplete(component, { ...args })
    case 'default':
    default:
      return {
        ...render(component),
      }
  }
}

export * from '@testing-library/react'
// export { default as userEvent } from '@testing-library/user-event'
export { customRender as render }
