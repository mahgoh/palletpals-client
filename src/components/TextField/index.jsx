import PropTypes from 'prop-types'
import { classNames } from '@/utils/common'

export default function TextField({
  name,
  type,
  label,
  value,
  onChange,
  placeholder,
  error,
  showError,
  className,
  ...props
}) {
  if (!name) {
    return null
  }

  return (
    <div>
      {label && (
        <label
          htmlFor={name}
          className={classNames(
            className,
            'mb-1 block text-sm font-medium text-gray-700 dark:text-gray-100'
          )}
        >
          {label}
        </label>
      )}
      <input
        id={name}
        name={name}
        type={type}
        className={classNames(
          className,
          error && 'border-red-500',
          'block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500'
        )}
        placeholder={placeholder}
        onChange={onChange}
        value={value}
        {...props}
      />
      {error && showError && (
        <p className="mt-2 text-sm text-red-500" id={`${name}-error`}>
          {error}
        </p>
      )}
    </div>
  )
}

TextField.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['text', 'email', 'password', 'number']),
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  error: PropTypes.string,
  showError: PropTypes.bool,
  className: PropTypes.string,
}

TextField.defaultProps = {
  type: 'text',
  label: null,
  value: '',
  onChange: () => {},
  placeholder: '',
  error: null,
  showError: true,
}
