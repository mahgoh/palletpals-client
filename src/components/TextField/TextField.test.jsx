import React from 'react'
import { render, screen } from '@/utils/test-utils'
import TextField from '.'

describe('textfield', () => {
  it('should render', () => {
    render(<TextField name="username" />)
    expect(screen.getByRole('textbox')).not.toBeFalsy()
    expect(screen.getByRole('textbox')).toMatchSnapshot()
  })

  it('should not render', () => {
    const { container } = render(<TextField />)
    expect(container).toBeEmptyDOMElement()
  })

  it('should render with label', () => {
    const { container } = render(<TextField name="username" label="Username" />)
    expect(container).toMatchSnapshot()
  })

  it('should render with error', () => {
    const { container } = render(
      <TextField name="username" label="Username" error="Required" />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with value', () => {
    const { container } = render(
      <TextField name="username" label="Username" value="test" />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with placeholder', () => {
    const { container } = render(
      <TextField name="username" label="Username" placeholder="Username" />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render with type', () => {
    const { container } = render(
      <TextField name="password" label="Password" type="password" />
    )
    expect(container).toMatchSnapshot()
  })
})
