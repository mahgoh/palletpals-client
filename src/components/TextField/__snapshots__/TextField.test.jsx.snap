// Vitest Snapshot v1

exports[`textfield > should render 1`] = `
<input
  class="block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
  id="username"
  name="username"
  placeholder=""
  type="text"
  value=""
/>
`;

exports[`textfield > should render with error 1`] = `
<div>
  <div>
    <label
      class="mb-1 block text-sm font-medium text-gray-700 dark:text-gray-100"
      for="username"
    >
      Username
    </label>
    <input
      class="border-red-500 block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
      id="username"
      name="username"
      placeholder=""
      type="text"
      value=""
    />
    <p
      class="mt-2 text-sm text-red-500"
      id="username-error"
    >
      Required
    </p>
  </div>
</div>
`;

exports[`textfield > should render with label 1`] = `
<div>
  <div>
    <label
      class="mb-1 block text-sm font-medium text-gray-700 dark:text-gray-100"
      for="username"
    >
      Username
    </label>
    <input
      class="block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
      id="username"
      name="username"
      placeholder=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`textfield > should render with placeholder 1`] = `
<div>
  <div>
    <label
      class="mb-1 block text-sm font-medium text-gray-700 dark:text-gray-100"
      for="username"
    >
      Username
    </label>
    <input
      class="block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
      id="username"
      name="username"
      placeholder="Username"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`textfield > should render with type 1`] = `
<div>
  <div>
    <label
      class="mb-1 block text-sm font-medium text-gray-700 dark:text-gray-100"
      for="password"
    >
      Password
    </label>
    <input
      class="block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
      id="password"
      name="password"
      placeholder=""
      type="password"
      value=""
    />
  </div>
</div>
`;

exports[`textfield > should render with value 1`] = `
<div>
  <div>
    <label
      class="mb-1 block text-sm font-medium text-gray-700 dark:text-gray-100"
      for="username"
    >
      Username
    </label>
    <input
      class="block w-full rounded-md border-gray-300 py-3 shadow-sm focus:border-orange-500 focus:ring-orange-500 dark:border-gray-700 dark:bg-gray-900 dark:focus:border-orange-500"
      id="username"
      name="username"
      placeholder=""
      type="text"
      value="test"
    />
  </div>
</div>
`;
